File: Queue and Stack/Stack/MyStack.java
/**
* Stack - A last-in-first-out date structure
* A dynamic array can be used to implement a stack
*/

import java.util.*;

public class MyStack {
	
	/**
	* Implementing a stack astract data type using a dynamic array
	* This is using a adapter pattern, where a instance of class List is adapted to define a custom stack class
	*/

	private List<Integer> data;

	public MyStack() {
		data = new ArrayList<Integer>();
	}

	/** Insert element into stack */
	public void push(int x) {
		data.add(x);
	}

	/** Check if stack is empty or not */
	public boolean isEmpty() {
		return data.isEmpty();
	}

	/** Get top element if stack */
	public int top() {
		if(!isEmpty()){
			return data.get(data.size() - 1);
		}
		return 0;
	}

	/** Delete element from stack, return true if operation is successful */
	public boolean pop() {
		if (isEmpty()) {
			return false;
		}
		data.remove(data.size() - 1);
		return true;
	}

	public static void main(String[] args) {

		MyStack stack = new MyStack();

		stack.push(1);
		System.out.println(stack.top());
		System.out.println(stack.pop());
		System.out.println(stack.top());
		System.out.println(stack.pop());

	}
}

File: Queue and Stack/Stack/MyStack.py
from Empty import Empty


class MyStack:

    """
    Implement a stack using python's list class
    This uses the adapter pattern where the list class is adapted to implement a custom stack class
    """

    def __init__(self):
        """ Create a empty stack """
        self._data = []

    def __len__(self):
        """ Return the number of elements in the stack"""
        return len(self._data)

    def is_empty(self):
        """Return True if stack is empty"""
        return len(self._data) == 0

    def push(self, e):
        """ Add element e to the top of the stack"""
        self._data.append(e)

    def top(self):
        """Return element at the top of the stack"""
        if self.is_empty():
            raise Empty("Stack is empty")
        return self._data[-1]

    def pop(self):
        """Remove and return the last element from the top of the stack"""
        if self.is_empty():
            raise Empty("Stack is empty")
        return self._data.pop()


if __name__ == '__main__':
    my_stack = MyStack()
    my_stack.push(1)
    print(my_stack.top())
    print(my_stack.pop())
    print(my_stack.pop())
